linera-rpc/src/grpc/client.rs

```diff
@@ -94,43 +94,77 @@ impl GrpcClient {
+    where
+        F: Fn(ValidatorNodeClient<transport::Channel>, Request<R>) -> FUT,
+        FUT: Future<Output = Result<tonic::Response<S>, Status>>,
+        R: IntoRequest<R>,
+    {
+        debug!(request = ?request, "sending gRPC request");
+        let mut retry_count = 0;
+        loop {
+            let request_inner = request
+                .clone()
+                .try_into()
+                .map_err(|_| NodeError::GrpcError {
+                    error: "could not convert request to proto".to_string(),
+                })?;
+            match f(self.client.clone(), Request::new(request_inner)).await {
+                Err(s) if Self::is_retryable(&s) && retry_count < self.max_retries => {
+                    let delay = self.retry_delay.saturating_mul(retry_count);
+                    retry_count += 1;
+                    linera_base::time::timer::sleep(delay).await;
+                    continue;
+                }
+                Err(s) => {
+                    return Err(NodeError::GrpcError {
+                        error: format!("remote request [{handler}] failed with status: {s:?}",),
+                    });
+                }
+                Ok(result) => return Ok(result.into_inner()),
+            };
+        }
+    }
+
+    #[allow(clippy::result_large_err)]
+    fn try_into_chain_info(
+        result: api::ChainInfoResult,
+    ) -> Result<linera_core::data_types::ChainInfoResponse, NodeError> {
+        let inner = result.inner.ok_or(NodeError::GrpcError {
+            error: "missing body from response".to_string(),
         })?;
-        let request = Request::new(request_inner);
-        match $self
-            .client
-            .clone()
-            .$handler(request)
-            .await
-            .map_err(|s| NodeError::GrpcError {
-                error: format!(
-                    "remote request [{}] failed with status: {:?}",
-                    stringify!($handler),
-                    s
-                ),
-            })?
-            .into_inner()
-            .inner
-            .ok_or(NodeError::GrpcError {
-                error: "missing body from response".to_string(),
-            })? {
+        match inner {
             Inner::ChainInfoResponse(response) => {
                 Ok(response.try_into().map_err(|err| NodeError::GrpcError {
                     error: format!("failed to marshal response: {}", err),
                 })?)
             }
             Inner::Error(error) => {
-                Err(bincode::deserialize(&error).map_err(|err| NodeError::GrpcError {
-                    error: format!("failed to marshal error message: {}", err),
-                })?)
+                Err(
+                    bincode::deserialize(&error).map_err(|err| NodeError::GrpcError {
+                        error: format!("failed to marshal error message: {}", err),
+                    })?,
+                )
             }
         }
+    }
+}
+
+macro_rules! client_delegate {
```

Comment:
Must this be a macro btw?

Reply:
It's just convenient to reduce boilerplate at the call sites. We could do away with it entirely, but it would probably add 6 lines per call site.